### PR TITLE
Drop support for Node 10, 12, 15, 17

### PR DESCRIPTION
**Pull request by aggmoulik**
- [x] Drop Node 10 Support ([end-of-life](https://nodejs.org/en/about/releases/) 2021-04)
- [x] Drop Node 12 support (end-of-life 2022-04-30)
- [x] Drop Node 15 Support (end-of-life 2021-06-01)
- [x] Drop Node 17 support (end-of-life 2022-06-01)
- [x] Remove `flat()` utility
- [x] Remove `flatMap()` utility

Part of v11 release (https://github.com/ember-cli/eslint-plugin-ember/issues/1169).

### CODE REVIEW

**Comment by bmish** [comment]
We'll also want to drop Node 15 support in this PR.

This PR can eventually be part of the v11 release (https://github.com/ember-cli/eslint-plugin-ember/issues/1169), but note that the v11 release is not urgent/imminent right now. Normally, we wait until we have additional recommended rules to enable before performing a major release.

File: README.md [hunk 1]
```diff
@@ -9,7 +9,7 @@
 ## ❗️Requirements
 
 - [ESLint](https://eslint.org/) `>= 6`
-- [Node.js](https://nodejs.org/) `10.* || 12.* || >= 14`
+- [Node.js](https://nodejs.org/) `12.* || >= 14`
```
Comment (bmish) [inline README.md:12]: Here's how we can drop Node 15 support (which is end-of-life already).

```suggestion
- [Node.js](https://nodejs.org/) `12.* || 14.* || >= 16`
```

File: package.json [hunk 2]
```diff
@@ -99,7 +99,7 @@
     "eslint": ">= 6"
   },
   "engines": {
-    "node": "10.* || 12.* || >= 14"
+    "node": "12.* || >= 14"
```
Comment (bmish) [inline package.json:102]: Here's how we can drop Node 15 support (which is end-of-life already).

```suggestion
    "node": "12.* || 14.* || >= 16"
```

https://nodejs.org/en/about/releases/

File: README.md [hunk 3]
```diff
@@ -9,7 +9,7 @@
 ## ❗️Requirements
 
 - [ESLint](https://eslint.org/) `>= 6`
-- [Node.js](https://nodejs.org/) `10.* || 12.* || >= 14`
+- [Node.js](https://nodejs.org/) `12.* || 14.* || >= 16`
```
Comment (bmish) [inline README.md:12]:
```suggestion
- [Node.js](https://nodejs.org/) `14.* || 16.* || >= 18`
```

File: package.json [hunk 4]
```diff
@@ -100,7 +100,7 @@
     "eslint": ">= 6"
   },
   "engines": {
-    "node": "10.* || 12.* || >= 14"
+    "node": "12.* || 14.* || >= 16"
```
Comment (bmish) [inline package.json:103]:
```suggestion
    "node": "14.* || 16.* || >= 18"
```

**Review by lin-ll — APPROVED** [review]

**Review by bmish — APPROVED** [review]
Thanks!

**Pull request merged**